SLE-Micro Container Host for ppc64le - fixes

m-dati opened the pull request:
Description:
- SLE-micro case added in `test_opensuse_based_image`;
- utils:`ensure_ca_certificates` missing CA uri-repos for sle-micro, management fixed;
- toolbox: when explicit uri needed, in place of CONTAINER_IMAGE_TO_TEST, introduced `CONTAINER_IMAGE_TO_TEST_EXTRA`, separation from BCI needed for that logic, otherwise causing next `get_image_uri` calls in SLE-M tests to override CONTAINER_IMAGE_VERSIONS.
- image::`get_image_uri` call, `distri` parameter defined;
- zypper_call: add retry input variable, to change the fixed 5 loops.

References:
- Related ticket: https://progress.opensuse.org/issues/162689
- Needles: N/A
- Verification run [other coming soon]:
  https://openqa.suse.de/tests/15649782
  https://openqa.suse.de/tests/15656966

m-dati commented:
If more commits or PR split needed, I'm open to suggestions.

m-dati commented on tests/containers/image.pm:
```diff
@@ -81,7 +81,7 @@ sub run {
     for my $version (split(/,/, $versions)) {
-        my $image = get_image_uri(version => $version);
+        my $image = get_image_uri(version => $version, distri => ($version =~ /-SP/ ? 'sle' : undef()));
 
```
This update is needed when HDD DISTRI is different from containers distri inCONTAINER_IMAGE_VERSIONS, like in this SLE-M cases. the routine get by default test's `DISTRI`

mloviska replied:
sle 16 does not have any `SP` that might be a future potential problem

m-dati replied:
Ops, I was not aware of such changes. But, this applies to `CONTAINER_IMAGE_VERSIONS` items only, like `15-SP2,15-SP3,15-SP4,...`: how should we expect that list with 16?

m-dati replied:
But eventually in that case, we'll add a more condition : `($version =~ /-SP|16/ ? ...`

grisu48 replied:
From what I see this is the only code change that is actually fixing something, I would make this alone a PR and split the other changes also into different PRs.

grisu48 replied:
How is this fixing something? Also, keep in mind that we also need to run this on openSUSE, so this might break stuff over there.

m-dati replied:
No, this is not the only code change.
Here is needed too: 
https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383/files#diff-8a38b5d8cac8d3a4ff4881bd952d410f3d640aacd3dedceea2fedea2095f46f9R185

m-dati replied:
and yes: it is TL;DR to explain, but update is tricky and needed, due to the BCI Container test flows

mloviska commented on lib/utils.pm:
```diff
@@ -588,6 +588,7 @@ sub zypper_call {
     my $timeout = $args{timeout} || 700;
     my $log = $args{log};
     my $dumb_term = $args{dumb_term} // is_serial_terminal;
+    my $retry = $args{retry} || 5;
```
`my $retry = $args{retry} // 5;` As `0` is a valid value IMO in case you simply want to have 0 retries.

m-dati replied:
0 reties means No-run at all, being the command `for 1 .. 0`: isn't it?
Also, the symbol `//` is used when `undefined`, so the $retry can even be empty and provide error

mloviska replied:
IMHO, the loop type is not correct. It should  always run at least once. If there is undef the retry will be the default value. In this implementation we cannot disable retry.

m-dati replied:
I left the previous old existing setting for default: 
if `$retry` is empty, then 5 assigned, as before it worked.

But also retry=0, here becomes 5

m-dati replied:
In last update: `undef`=5; `assigned but '' or 0`=1; `>0`=$retry.

m-dati replied:
In last update I unified the settings, adding a description:
`# (retry >=0 or ='': unchanged; undef: 5) result ='' or 0: 1`

mloviska replied:
The base of the problem is following:

- user wants to run a `X_retry` so it means it will always run at least once
- number of retries can be parametrized, that is nice, but it should lead to `1 + retries` runs

In this implementation the caller would think that there are 5 retries at least, but there are only 4 in fact.

mloviska replied:
IMO this would simply the code and it would be also cleaner, straightforward and maybe more safe

m-dati replied:
It is a matter of definitions: _retries after `first` attempt failed_ or _total runs while failing_ (we could rename variable).
In this implementation: 1 means 1 run only total (no retry); 2 means 1 retry if failed or 2 total runs whil failing, etc...
To well-use our routines we have the heading routine descipition: the user should clarify all logic and usage there: I will update it too, when behavior clarified.

mloviska replied:
I have a feeling that this PR is diverging into all possible spaces at once. I would be happy to see that `zypper_call` is refactored/updated/you name it in a separate PR, slow track that all users can drop their insights and thoughts of to make the code better. I do not see any strong relation between `ppc` and `zypper_call` that is required at the moment. It would be beneficial for everyone if `zypper_call` improvements would be separated and that you can improve that code (which would be welcomed) in a different PR :)

m-dati replied:
Yes I already mentioned it before in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383#issuecomment-2407440264.
A weak correlation and motivation that led me to this change is the introduction of sure failing check on sle-micro CA check: original loop is 5 retry, I attempting to reduce it to 3 at most.
But if it is a problem, I will revert this change and move it to a different next PR.

mloviska replied:
The problem is that cert URL seems to be wrongly composed. If that is fixed, all the stuff should run smoothly as it runs well for other archs

mloviska commented on lib/utils.pm:
```diff
@@ -2374,11 +2375,21 @@ sub ensure_ca_certificates_suse_installed {
+            if (is_sle_micro) {
+                transactional::trup_call('--continue run update-ca-certificates -v');
+            } else {
+                my $res = script_run 'update-ca-certificates -v';
```
why not `assert_script_run`?

m-dati replied:
ok

m-dati replied:
done

mloviska replied:
Now it is `$res` + `assert_script_run` which dont play well together. My point in general was that passing the exit code might not be beneficial

m-dati replied:
I agree and this topic is discussed in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383#discussion_r1799012287: depending on the final solution, script_run with ouptut or assert_... will be used accordingly.

mloviska commented on tests/microos/toolbox.pm:
```diff
@@ -60,7 +60,7 @@ sub run {
     $self->create_user;
 
-    my $toolbox_image_to_test = get_var('CONTAINER_IMAGE_TO_TEST');
+    my $toolbox_image_to_test = get_var('CONTAINER_IMAGE_TO_TEST_EXTRA');
```
Why this got renamed?

m-dati replied:
I realized that, in this module for toolbox tests, when needed to update the default uri, we use that same parameter also used in other containers modules where logic is different, like image_podman, where failed in a VR.

In fact, in the the loop above, `tests/containers/image.pm:L#84`, calling [get_image_uri](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/633f61d452c908689aea88219bbb6b48c1440a17/lib/containers/urls.pm#L206), it will soon return that assigned url, instead of continuing on distri/version found in `CONTAINER_IMAGE_VERSIONS`. That url is not what we expect in containers tests.

m-dati replied:
VR 1 above: 
[Here in toolbox](https://openqa.suse.de/tests/15649782#step/toolbox/12) we use `..._EXTRA`;
[here in image_podman](https://openqa.suse.de/tests/15649782#step/image_podman/115), using the standard parameter, the loop can inspect the containers in CONTAINER_IMAGE_VERSIONS: but still we can set eventual `...TO_TEST` for that logic, nothing to do with toolbox.

This happens in particular when HDD_ and elements in CONTAINER_IMAGE_VERSIONS have _diffrent distros_, like for SLE-M and SLES.

grisu48 replied:
This variable has been context-sensitive (i.e. test-run sensitive) since years and it worked just fine. The name `CONTAINER_IMAGE_TO_TEST` is descriptive to what it is, it defines the container under test.

We also don't use different `ISO` or `HDD_` variables depending on the test run, this is a more generic setting that served us well in the past.

***

When changing something this fundamental it needs to be well motivated. So please @m-dati, either make your case or please let's not rename something that is used all over the place since a long time without the need for doing so.

m-dati replied:
Please, before read carfully the description provided in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383#discussion_r1797118768.

I am not renaming anything, but **using another parameter** to address the local toolbox need. The original parameter CONTAINER_IMAGE_TO_TEST scope and usage reamain the same.

I simply realized that the `image explicit definition` in this toolbox module also affecting another next module, where the original CONTAINER_IMAGE_TO_TEST used for a totally different meaning and scope.

Therefore, this means that this for toolbox is a **totally different parameter**.
I named it _similar to the original_ simply adding a variation (_EXTRA), to mantain its origin, but to generalize it for further usage, but we can opt for other naming, like e.g. TOOLBOX_IMAGE_TO_TEST

grisu48 replied:
This point requires a discussion, let's reserve a time slot on Thursday for it.

rbmarliere commented on lib/utils.pm:
```diff
@@ -2371,16 +2372,29 @@ This functions checks if ca-certificates-suse is installed and if it is not it a
-        if (is_sle_micro) {
-            transactional::trup_call('--continue pkg install ca-certificates-suse');
+                                                                             # exit codes defined will not stop on error.
+        my $res = zypper_call("ar --refresh http://download.suse.de/ibs/SUSE:/CA/SLE_$distversion/SUSE:CA.repo", retry => 3, exitcode => [0, 1, 2, 3, 4]);
```
Using both $res and $ret hinders readability, in my opinion. Since $res is only used once, what do you think about moving the zypper_call into the if statement?

m-dati replied:
the zypper exit code is the basilar value for block triggering and moving that command into the condition is much more readability hindering.
I agree that variables names are too similar, I will rename `res`.

m-dati replied:
I renamed a variable.
Also, let's image that long zypper command into the `if` matching 0: on that right side of the screen we will lost :slightly_smiling_face:

rbmarliere commented on lib/utils.pm:
```diff
@@ -600,7 +601,7 @@ sub zypper_call {
                     /var/log/zypper.log
                     ';
-    for (1 .. 5) {
+    for (1 .. ($retry or 1)) {
```
Will this ever evaluate to 1 .. 1? Doesn't $retry default to 5?

m-dati replied:
see https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383#discussion_r1797238569

m-dati commented:
Here are other VRs after last update:

- sle-micro-6.1-Base-ppc-4096-ppc64le-Build21.3-slem_podman@ppc64le-emu - 
https://openqa.suse.de/tests/15660596  

- sle-micro-6.1-Base-ppc-512-SelfInstall-ppc64le-Build21.3-slem_docker@ppc64le-emu - 
https://openqa.suse.de/tests/15660597  

- sle-micro-6.1-Default-ppc-512-ppc64le-Build21.3-slem_podman@ppc64le-emu - 
https://openqa.suse.de/tests/15660598  

Non regression:

- sle-micro-6.1-Base-RT-encrypted-x86_64-Build21.3-slem_podman@uefi - 
https://openqa.suse.de/tests/15660599  PASS 

- sle-micro-6.1-Base-RT-SelfInstall-x86_64-Build21.3-slem_fips_docker@64bit - 
https://openqa.suse.de/tests/15660600 PASS 

- sle-micro-6.1-Container-Image-Updates-x86_64-Build6.1_3.14-sle_micro_toolbox_image@64bit - 
https://openqa.suse.de/tests/15660593 PASS

- sle-micro-6.1-Container-Image-Updates-x86_64-Build6.1_3.14-sle_micro_toolbox_image@64bit - 
https://openqa.suse.de/tests/15660594 PASS

- sle-15-SP6-BCI-Updates-x86_64-Build26.3_micro-image-bci-micro_15.6_on_SLES_15-SP5_podman@64bit - 
https://openqa.suse.de/tests/15660592 PASS

mloviska commented on lib/utils.pm:
```diff
@@ -2371,16 +2372,30 @@ This functions checks if ca-certificates-suse is installed and if it is not it a
+            }
         }
     }
+    return $ret;
```
I am not sure whether returning the caller a int makes it any better. By the name the function should either succeed or die.

m-dati replied:
Ok, but let we consider some new situation raised, due to  [IBS CA](http://download.suse.de/ibs/SUSE:/CA/), where NO sle-micro present: because of that, I introduced the `out` : `true` for SLE(S)/OpenSUSE OR `fail` (SLE-Micro or else) and result became a 4 way output.
IMHO the decision hould be passed from caller by  a more option  `continue_on_fail` or die: 
wdyt ?

mloviska replied:
until 6.0, sle-micro versions were mapped to sle, which is the previous code reflecting. The question is how we should continue over here and that is a question on PjM rather than any QA engineer can answer.

m-dati replied:
Please check last commit.

mloviska replied:
I think I have found the cert folder, or at least it might be worth to be tested `/SLE-Factory/` in case of slem 6.0+

m-dati replied:
I already noted it, but I had no info on if-and-when it can be used, in place of which case.

mloviska replied:
> I already noted it, but I had no info on if-and-when it can be used, in place of which case.

First of all I would try to test it, then you can come to the project channel with direct questions. We are not building the assets so we cannot know :)

grisu48 commented on lib/utils.pm:
```diff
@@ -2371,16 +2372,30 @@ This functions checks if ca-certificates-suse is installed and if it is not it a
+
+        # exit codes defined will not stop on error.
+        my $out = zypper_call("ar --refresh http://download.suse.de/ibs/SUSE:/CA/SLE_$distversion/SUSE:CA.repo", retry => 3, exitcode => [0, 1, 2, 3, 4]);
+        if ($out == 0) {
```
I really don't like it that we're trying to workaround infra issues here. This makes the test flow unpredictable.

Can you justify why this is needed? We use this all over the place and I'm not aware on infra issues on such a scale that would justify this addition.

grisu48 commented on lib/utils.pm:
```diff
@@ -588,7 +588,8 @@ sub zypper_call {
     my $dumb_term = $args{dumb_term} // is_serial_terminal;
-
+    # (retry >=0 or ='': unchanged; if undef: 5) would first term be '' or 0: 1
+    my $retry = ($args{retry} // 5) or 1;
```
And why do we make this change in a fixing PR?

grisu48 reviewed (changes requested):
I see several problems with this PR:

1. For a fixing PR it does too many improvements. A fixing PR should only do the required fixes
2. The PR touches some problematic places, that need to have some discussions. I would split them into separate PR to keep them focused
3. It does too many things at once. I would recommend to split every logical piece into an individual PR, which makes the review easier and doesn't block the whole PR but only sub-pieces.

@m-dati I would recommend to split the individual pieces into different PRs and proceed there. This PR will be difficult to merge, because it does too many things at once and some of them require discussions.

m-dati commented on lib/containers/container_images.pm:
```diff
@@ -182,6 +182,8 @@ sub test_opensuse_based_image {
             record_info "non-SLE host", "This host ($host_id) does not support zypper service";
         }
+    } elsif ($image_id =~ /^(sl|sle)-micro$/) {
+        validate_script_output qq{$runtime container run --entrypoint '/bin/bash' --rm $image -c 'cat /etc/os-release'}, sub { /PRETTY_NAME="SUSE Linux Micro .*"/ };
```
It used also for fo OSD tests [here](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/9af8a47087054f20ee37219e9ce68e624b1f5934/tests/containers/image.pm#L99) and it impacted the new SLE-M tests: this update is needed for the actual structure.

grisu48 replied:
openSUSE based images cannot contain SLE-Micro in their `/etc/os-release`. If this is the case, the test design is broken and then we need to fix that.

m-dati replied:
This routine is used ALSO for SLES, as by the code `if ($image_id =~ 'sles')...`; therefore, is not that strange adding here SLE-Micro too. Its name is misleading

m-dati commented:
Second issue I encountered when I added ppc64le for SLE-M 6.1 in this poo's test group, after boot ok, was related to missing certificates on the new image, preventing to pull from https registry, in podman tests.
But then also resulted missing on IBS the SLE-micro certificate folder.
Therefore I updated the [code](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383/files#diff-47a0a7f15c1bd3166c2ca1c8c3f4f0c7f7beb1b1b5c4d349e4da8cab17e9ae5eR2373), to get it from ca.suse.de when no repo found.

Now, after some discussion on it, given last [concepts](https://suse.slack.com/archives/C02CBB35W5B/p1728924896947509) and [suggestion](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/20383#discussion_r1799592446), I will provide a simpler update.

I go splitting main changes of this PR in different sub-PRs.

m-dati commented:
In other PR-s some more discussions addressed actual test flow status and needed changes, then applied.
For the related p.o.o. ticket the work is done, then this PR can be _closed_.
